File: C++/00. LeetCode/TestProgram.cpp
#include "TestProgram.h"
#include "Printout.h"

namespace LeetCode
{
	TestProgramBase::TestProgramBase(ostream* out, PrintoutMode mode) :
		out(out), mode(mode), needEOL(false)
	{
		Reset();
	}

	void TestProgramBase::SetOutput(ostream* out, PrintoutMode mode)
	{
		this->out = out;
		this->mode = mode;
	}

	void TestProgramBase::PrintSummary(bool reset)
	{
		if (out != nullptr) {
			Newline();
			*out << "Total " << sumTests << " tests, " << sumPassed << " passed, " << sumFailed << " failed." << endl;
		}
		if (reset) {
			Reset();
		}
	}

	void TestProgramBase::Reset()
	{
		setTests = setPassed = setFailed = 
		sumTests = sumPassed = sumFailed = 0;
		setTime = sumTime = duration_t::zero();
	}

	void TestProgramBase::Newline()
	{
		if (needEOL) {
			*out << endl;
			needEOL = false;
		}
	}

	void TestProgramBase::Backspace(int count)
	{
		if (needEOL) {
			for (int i = 0; i < count; i++) {
				*out << '\b';
			}
			for (int i = 0; i < count; i++) {
				*out << ' ';
			}
			for (int i = 0; i < count; i++) {
				*out << '\b';
			}
		}
	}


	template<class TSample, typename TResult>
	TestProgram<TSample, TResult>::TestProgram(ostream* out, PrintoutMode mode, Test_t test, Verify_t verify) :
		TestProgramBase(out, mode),
		test(test), verify(verify)
	{
	}

	template<class TSample, typename TResult>
	void TestProgram<TSample, TResult>::Test(Sample_t& sample, bool printSample, bool printFailed)
	{
		clock_t::time_point start = clock_t::now();
		test(sample);
		clock_t::time_point end = clock_t::now();
		duration_t time = end - start;

		setTests++;
		sumTests++;
		setTime += time;
		sumTime += time;

		bool success, unknown = false;
		if (verify != nullptr) {
			success = verify(sample);
		} else if (sample.Expected() != nullptr) {
			success = (*sample.Result() == *sample.Expected());
		} else {
			unknown = true;
		}

		if (unknown) {
			sample.SetStatus(TestStatus::Unknown);
		}
		else if (success) {
			sample.SetStatus(TestStatus::Passed);
			setPassed++;
			sumPassed++;
		} else {
			sample.SetStatus(TestStatus::Failed);
			setFailed++;
			sumFailed++;
			if (printFailed)
				printSample = true;
		}

		if (printSample) {
			PrintSample(sample);
		}
	}

	template<class TSample, typename TResult>
	void TestProgram<TSample, TResult>::RunTests(vector<Sample_t>& samples, bool printSamples, bool printFailed)
	{
		setTests = setPassed = setFailed = 0;
		setTime = duration_t::zero();

		for (auto it = samples.begin(); it != samples.end(); it++)
		{
			Sample_t& sample = *it;
			Test(sample, printSamples, printFailed);
		}
	}

	template<class TSample, typename TResult>
	void TestProgram<TSample, TResult>::RunTests(Samples_t& samples, bool printSamples, bool printFailed)
	{
		setTests = setPassed = setFailed = 0;
		setTime = duration_t::zero();

		for (auto it = samples.begin(); it != samples.end(); it++)
		{
			Sample_t* sample = *it;
			if (sample == nullptr)
				break;
 
			Test(*sample, printSamples, printFailed);
		}
	}

	template<class TSample, typename TResult>
	void TestProgram<TSample, TResult>::RandomTests(Random_t& random, int nSamples, int N)
	{
		if (out != nullptr) {
			*out << "Random tests";
			if (N > 0) {
				*out << " for N=" << N;
			}
			*out << ": Generating...";
			needEOL = true;
		}

		random.PrepareSet(nSamples, N);
		nSamples = random.NumSamples();
		N = random.N();

		Samples_t samples(nSamples);
		for (int i = 0; i < nSamples; i++)
		{
			Sample_t* sample = random.GetSample();
			if (sample == nullptr)
				break;

			samples[i] = sample;
		}


		if (out != nullptr) {
			Backspace(13);
			*out << "Running...";
		}
		RunTests(samples, false);

		if (out != nullptr) {
			chrono::milliseconds ms = chrono:: duration_cast<chrono::milliseconds>(setTime);
			Backspace(10);
			*out << ms.count() << "ms";
			if (setPassed > 0 || setFailed > 0) {
				*out << ", " << setPassed << " passed " << setFailed << " failed";
			}
			*out << endl;
			needEOL = false;
		}
	}

	template<class TSample, typename TResult>
	void TestProgram<TSample, TResult>::RandomTests(Random_t& random, int nSamples, int minN, int maxN, int stepN, int multN)
	{
		if (maxN < minN)
			maxN = minN;

		if (multN <= 1) {
			multN = 1;
			if (stepN < 1)
				stepN = 1;
		}
	
		for (int N = minN; N <= maxN; N = N * multN + stepN)
		{
			RandomTests(random, nSamples, N);
		}
	}


	template<class TSample, typename TResult>
	void TestProgram<TSample, TResult>::PrintSample(const Sample_t& sample)
	{
		if (out == nullptr)
			return;

		Newline();

		const TSample::Args_t& args = sample.Args();
		*out <<  Printout<TSample::Args_t>(args, mode);;
		needEOL = true;

		if (sample.Result() != nullptr) {
			*out << " => " << Printout<TResult>(*sample.Result(), mode);
			needEOL = true;
		}
		if (sample.Status() == TestStatus::Passed) {
			*out << "  Passed";
			needEOL = true;
		}
		else if (sample.Status() == TestStatus::Failed) {
			*out << "  Failed";
			if (sample.Expected() != nullptr) {
				*out << ", expected " << Printout<TResult>(*sample.Expected(), mode);
			}
			needEOL = true;
		}

		Newline();
	}


#define TEST_TYPE1(TResult, TArg1)\
	template class TestProgram<TestSample<TResult, TArg1>, TResult>;

#define TEST_TYPE2(TResult, TArg1, TArg2)\
	template class TestProgram<TestSample<TResult, TArg1, TArg2>, TResult>;

#define TEST_TYPE3(TResult, TArg1, TArg2, TArg3)\
	template class TestProgram<TestSample<TResult, TArg1, TArg2, TArg3>, TResult>;

#include "TestTypes.h"

} //namespace LeetCode

File: Java/00. LeetCode/src/LeetCode/TestProgram.java
/**
 * 
 */
package LeetCode;

import java.util.function.Consumer;
import java.util.function.Function;
import java.io.PrintStream;


/**
 * @author Roman Dementiev
 *
 */

public abstract class TestProgram<TSample extends TestSample<?,?,?,?>>
{
	protected PrintStream out;
	protected Consumer<TSample> test = null;
	protected Function<TSample, TestStatus> verify = null;
	protected long setTime = 0, sumTime = 0;
	protected int setTests = 0, setPassed = 0, setFailed = 0;
	protected int sumTests = 0, sumPassed = 0, sumFailed = 0;
	protected boolean needEOL = false;
	
	
	public TestProgram()
	{
		this(System.out);
	}
	
	public TestProgram(PrintStream out)
	{
		this.out = out;
	}
	
	public void reset(boolean resetSummary)
	{
		setTime = setTests = setPassed = setFailed = 0;
		if (resetSummary) {
			sumTime = sumTests = sumPassed = sumFailed = 0;
		}
	}
	
	public void setTest(Consumer<TSample> test)
	{
		this.test = test;
	}
	
	public void setVerify(Function<TSample, TestStatus> verify)
	{
		this.verify = verify;
	}
	
	protected void test(TSample sample, boolean printSample, boolean printFailed)
	{
		// TODO: Timing
		long start = System.nanoTime();
		test.accept(sample);
		long end = System.nanoTime();
		long time = end - start;

		setTests++;
		sumTests++;
		setTime += time;
		sumTime += time;
		
		TestStatus status;
		if (verify != null) {
			status = verify.apply(sample);
		} else {
			status = sample.verify();
		}
		sample.setStatus(status);
		
		if (status == TestStatus.Passed) {
			setPassed++;
			sumPassed++;
		} else if (status == TestStatus.Failed)  {
			setFailed++;
			sumFailed++;
			if (printFailed)
				printSample = true;
		}

		if (printSample && out != null) {
			out.println(sample.toString());
		}
	}

	public void runTests(TSample[] samples)
	{
		runTests(samples, true, true);
	}
	
	public void runTests(TSample[] samples, boolean printSample, boolean printFailed)
	{
		reset(false);
		for (TSample sample : samples) {
			if (sample == null)
				break;
			test(sample, printSample, printFailed);
		}
	}	

	public void randomTests(RandomSamples<TSample> randoms, TSample[] samples)
	{
		randomTests(randoms, samples, 0);
	}
	
	public void randomTests(RandomSamples<TSample> randoms, TSample[] samples, int N)
	{
		if (out != null) {
			out.print("Random tests");
			if (N > 0) {
				out.format(" for N=%d:", N);
			} else {
				out.print(" for random N:");
			}
			out.flush();
			needEOL = true;
		}

		samples = randoms.prepareSet(samples, samples.length, N);
		runTests(samples, false, true);

		if (out != null) {
			long ms = setTime / 1000000;
			out.format(" %dms", ms);
			if (setPassed > 0 || setFailed > 0) {
				out.format(", %d passed, %d failed", setPassed, setFailed);
			}
			out.println("");
			needEOL = false;
		}
	}

	public void randomTests(RandomSamples<TSample> randoms, TSample[] samples, int minN, int maxN, int stepN, int multN)
	{
		if (maxN < minN)
			maxN = minN;

		if (multN <= 1) {
			multN = 1;
			if (stepN < 1)
				stepN = 1;
		}
	
		for (int N = minN; N <= maxN; N = N * multN + stepN)
		{
			randomTests(randoms, samples, N);
		}
	}

	public void randomTests(RandomSamples<TSample> randoms, TSample[] samples, int minN, int maxN, int stepN)
	{
		randomTests(randoms, samples, minN, maxN, stepN, 1);
	}
	
	public void randomTests(RandomSamples<TSample> randoms, TSample[] samples, int minN, int maxN)
	{
		randomTests(randoms, samples, minN, maxN, 1, 1);
	}
	
	public void printSample(TSample sample)
	{
		if (out == null)
			return;
		
		newline();
		out.println(sample.toString());
	}
	
	public void printSummary(boolean reset)
	{
		if (out != null) {
			newline();
			out.print("Total ");
			out.print(sumTests);
			out.print(" tests, ");
			out.print(sumPassed);
			out.print(" passed, ");
			out.print(sumFailed);
			out.println(" failed.");
		}
		if (reset) {
			reset(true);
		}
	}

	protected void newline()
	{
		if (needEOL) {
			out.println("");
			needEOL = false;
		}
	}
};
